<a href="https://colab.research.google.com/github/Bashnalytics/circle363_holiday_challenge/blob/main/Laptop_Prices_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Building Fast Queries: CSV**

**Data Exploration**

In [30]:
import csv
import pandas as pd
df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
header = list(df)
rows = df.values.tolist()
print(header)
print(rows[:5])
print(len(rows))

['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']
[[1, 'Apple', 'MacBook Pro', 'Ultrabook', 13.3, 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', 1339.69], [2, 'Apple', 'Macbook Air', 'Ultrabook', 13.3, '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', 898.94], [3, 'HP', '250 G6', 'Notebook', 15.6, 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575.0], [4, 'Apple', 'MacBook Pro', 'Ultrabook', 15.4, 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', 2537.45], [5, 'Apple', 'MacBook Pro', 'Ultrabook', 13.3, 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB SSD', 'Intel Iris Plus Graphics 65

**Creating an Inventory Class**

In [31]:
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])

inventory = Inventory()
print(inventory.header)
print(len(inventory.rows))            

['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']
1303


**Finding a Laptop using the ID**

In [32]:
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])

  def get_laptop_from_id(self, laptop_id):
    for row in self.rows:
      if row[0] == laptop_id:
        return row
        return None                            

In [33]:
inventory = Inventory()
print(inventory.get_laptop_from_id(131))
print(inventory.get_laptop_from_id(1400))

[131, 'Dell', 'Inspiron 5770', 'Notebook', 17.3, 'Full HD 1920x1080', 'Intel Core i7 8550U 1.8GHz', '16GB', '256GB SSD +  2TB HDD', 'AMD Radeon 530', 'Windows 10', '2.8kg', 1396]
None


**Improving the ID Lookup Query**

In [34]:
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])
      self.id_to_row = {}
      for row in self.rows:
        self.id_to_row[row[0]] = row 

  def get_laptop_from_id(self, laptop_id):
    for row in self.rows:
      if row[0] == laptop_id:
        return row
        return None    

  def get_laptop_from_id_fast(self, laptop_id):
    if laptop_id in self.id_to_row:
      return self.id_to_row[laptop_id]
      return None    

In [35]:
inventory = Inventory()
print(inventory.get_laptop_from_id_fast(131))
print(inventory.get_laptop_from_id_fast(1400))

[131, 'Dell', 'Inspiron 5770', 'Notebook', 17.3, 'Full HD 1920x1080', 'Intel Core i7 8550U 1.8GHz', '16GB', '256GB SSD +  2TB HDD', 'AMD Radeon 530', 'Windows 10', '2.8kg', 1396]
None


**Comparing the Performance of the two methods (get_laptop_from_id and get_laptop_from_id_fast)**

In [37]:
import time
import random

ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]

inventory = Inventory()

total_time_no_dict = 0 
for identifier in ids:
  start = time.time()
  inventory.get_laptop_from_id(identifier)
  end = time.time()
  total_time_no_dict += end - start 

total_time_dict = 0 
for identifier in ids:
  start = time.time()
  inventory.get_laptop_from_id_fast(identifier)
  end = time.time()
  total_time_dict += end - start

print(total_time_no_dict)             
print(total_time_dict)

1.1169447898864746
0.008577108383178711


**Interpretation**

We can see a significant improvement in the performance of the second method. If we divide 1.116945 by 0.008577, we see that the new method is about 130 times faster for this input size.

**Two Laptop Promotion**

Write a method that finds whether we can spend a given amount of money by purchasing either one or two laptops.

In [39]:
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])
      self.id_to_row = {}
      for row in self.rows:
        self.id_to_row[row[0]] = row 

  def get_laptop_from_id(self, laptop_id):
    for row in self.rows:
      if row[0] == laptop_id:
        return row
        return None    

  def get_laptop_from_id_fast(self, laptop_id):
    if laptop_id in self.id_to_row:
      return self.id_to_row[laptop_id]
      return None   

  def check_promotion_dollars(self, dollars):
    for row in self.rows:
      if row[-1] == dollars:
        return True
        for row1 in self.rows:
          for row2 in self.rows:
            if row1[-1] + row2[-1] == dollars:
              return True
        return False     

In [41]:
inventory = Inventory()
print(inventory.check_promotion_dollars(1000))
print(inventory.check_promotion_dollars(442))

True
None


Optimizing the Two Laptop Promotion method


In [42]:
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])
    self.id_to_row = {}
    for row in self.rows:
      self.id_to_row[row[0]] = row 
    self.prices = set()
    for row in self.rows:
      self.prices.add(row[-1])

  def get_laptop_from_id(self, laptop_id):
    for row in self.rows:
      if row[0] == laptop_id:
        return row
        return None    

  def get_laptop_from_id_fast(self, laptop_id):
    if laptop_id in self.id_to_row:
      return self.id_to_row[laptop_id]
      return None   

  def check_promotion_dollars(self, dollars):
    for row in self.rows:
      if row[-1] == dollars:
        return True
        for row1 in self.rows:
          for row2 in self.rows:
            if row1[-1] + row2[-1] == dollars:
              return True
        return False   

  def check_promotion_dollars_fast(self, dollars):
    if dollars in self.prices:
      return True
      for price in self.prices:
        if dollars - price in self.prices:
          return True
      return False  

In [43]:
inventory = Inventory()
print(inventory.check_promotion_dollars_fast(1000))
print(inventory.check_promotion_dollars_fast(442))

True
None


**Comparing the Performance of the two methods (check_promotion_dollars and check_promotion_dollars_fast)**

In [45]:
prices = [random.randint(100, 5000) for _ in range(100)]

inventory = Inventory()  

total_time_no_set = 0 
for price in prices:
  start = time.time()
  inventory.check_promotion_dollars(price)
  end = time.time()
  total_time_no_set += end - start

total_time_set = 0
for price in prices:
  start = time.time()
  inventory.check_promotion_dollars_fast(price)
  end = time.time()
  total_time_set += end - start 
    
print(total_time_no_set)
print(total_time_set)

0.007682323455810547
0.00010323524475097656


**Interpretation**

We can see a significant improvement in the performance of the second method. If we divide 0.007682 by 0.000103, we see that the new method is about 75 times faster for this input size.

**Finding Laptops Within a Budget**

In [51]:
def row_price(row):
    return row[-1]
class Inventory():
  def __init__(self): 
    df = pd.read_csv("/content/laptop_price.csv", encoding='latin-1') 
    self.header = list(df)
    self.rows = df.values.tolist()
    for row in self.rows:
      row[-1] = int(row[-1])
    self.id_to_row = {}
    for row in self.rows:
      self.id_to_row[row[0]] = row 
    self.prices = set()
    for row in self.rows:
      self.prices.add(row[-1])
    self.rows_by_price = sorted(self.rows, key=row_price)

  def get_laptop_from_id(self, laptop_id):
    for row in self.rows:
      if row[0] == laptop_id:
        return row
        return None    

  def get_laptop_from_id_fast(self, laptop_id):
    if laptop_id in self.id_to_row:
      return self.id_to_row[laptop_id]
      return None   

  def check_promotion_dollars(self, dollars):
    for row in self.rows:
      if row[-1] == dollars:
        return True
        for row1 in self.rows:
          for row2 in self.rows:
            if row1[-1] + row2[-1] == dollars:
              return True
        return False   

  def check_promotion_dollars_fast(self, dollars):
    if dollars in self.prices:
      return True
      for price in self.prices:
        if dollars - price in self.prices:
          return True
      return False  
  def find_laptop_with_price(self, target_price):
    range_start = 0                                   
    range_end = len(self.rows_by_price) - 1                       
    while range_start < range_end:
      range_middle = (range_end + range_start) // 2  
      value = self.rows_by_price[range_middle][-1]
      if value == target_price:
        return range_middle                        
      elif value < target_price:
        range_start = range_middle + 1             
      else:                                          
        range_end = range_middle - 1
    if self.rows_by_price[range_start][-1] != target_price:
      return -1                                      
    return range_start
    
  def find_first_laptop_more_expensive(self, target_price):
      range_start = 0                                   
      range_end = len(self.rows_by_price) - 1                   
      while range_start < range_end:
        range_middle = (range_end + range_start) // 2  
        price = self.rows_by_price[range_middle][-1]
        if price > target_price:
          range_end = range_middle
        else:
          range_start = range_middle + 1
      if self.rows_by_price[range_start][-1] <= target_price:
        return -1                                   
      return range_start
  

In [52]:
inventory = Inventory()
print(inventory.find_first_laptop_more_expensive(1000))
print(inventory.find_first_laptop_more_expensive(10000))

683
-1
